File: pdatocfg.cpp
#include <iostream>
#include <sstream>
#include <vector>

#define breakline() cout << "-------------------------------------------------------------------------\n"
using namespace std;

void generate_permutations(vector<string> states, vector<string> &temp, vector<vector<string> > &permutes, int k) {
    if (k == 0) {
        permutes.push_back(temp);
    } else {
        for (int i = 0; i < states.size(); i++) {
            temp[k - 1] = states[i];
            generate_permutations(states, temp, permutes, k - 1);
        }
    }
}

void print_permutes(string from_state, string input, string top_stack, string to_state, string push_stack, vector<string> states) {
    int i, j;
    vector<string> temp(push_stack.size());
    vector<vector<string> > permutes;

    generate_permutations(states, temp, permutes, push_stack.size());

    for (i = 0; i < permutes.size(); i++) {
        cout << "[" << from_state << top_stack;
        cout << permutes[i][0] << "] -> " << input << " [" << to_state << push_stack[0];
        for (j = 1; j < permutes[i].size(); j++) {
            cout << permutes[i][j] << "] [" << permutes[i][j] << push_stack[j];
        }
        cout << permutes[i][0] << "]\n";
    }
}

void PDA_to_CFG(vector<string> &from_state, vector<string> &input, vector<string> &top_stack, vector<string> &to_state, vector<string> &push_stack, vector<string> states) {
    int i, j, n = from_state.size();
    cout << "Start State = S -> ";
    cout << "[" << states[0] << "Z" << states[0] << "]";
    for (i = 1; i < states.size(); i++) {
        cout << " | [" << states[0] << "Z" << states[i] << "]";
    }
    cout << endl;

    for (i = 0; i < n; i++) {
        cout << "\nδ(" << from_state[i] << "," << input[i] << "," << top_stack[i] << ") = (" << to_state[i] << "," << push_stack[i] << ")\n";
        if (push_stack[i].compare("ε") == 0) {
            cout << "[" << from_state[i] << top_stack[i] << to_state[i] << "] -> " << input[i] << endl;
        } else {
            print_permutes(from_state[i], input[i], top_stack[i], to_state[i], push_stack[i], states);
        }
    }

    cout << endl;
    for (i = 0; i < states.size(); i++) {
        for (j = 0; j < states.size(); j++) {
            cout << "[" << states[i] << "Z" << states[j] << "] -> ε\n";
        }
    }
}

int main() {
    vector<string> from_state;
    vector<string> to_state;
    vector<string> top_stack;
    vector<string> input;
    vector<string> push_stack;
    vector<string> states;

    int i, j;
    vector<string> temp;
    string line, ch, t;

    cout << "Enter states of the PDA separated by spaces: ";
    getline(cin, line);
    istringstream ss(line);
    while (ss >> ch) {
        states.push_back(ch);
    }

    cout << "Enter PDA transitions in the format δ(q,0,Z) = (q,XZ) (press Enter after each transition, type 'done' to finish):\n";

    while (true) {
        cout << "Transition: ";
        getline(cin, line);
        if (line == "done") break;

        istringstream ss(line);

        ss >> ch;
        istringstream s2(ch);
        i = 0;
        while (getline(s2, t, ',')) {
            if (i == 0) {
                t.erase(t.begin(), t.begin() + 3);
                from_state.push_back(t);
            } else if (i == 1) {
                input.push_back(t);
            } else if (i == 2) {
                t.erase(t.end() - 1, t.end());
                top_stack.push_back(t);
            }
            i++;
        }

        ss >> ch;
        ss >> ch;
        istringstream s3(ch);
        i = 0;
        while (getline(s3, t, ',')) {
            if (i == 0) {

                t.erase(t.begin(), t.begin() + 1);
                to_state.push_back(t);
            } else if (i == 1) {
                t.erase(t.end() - 1, t.end());
                push_stack.push_back(t);
            }
            i++;
        }
    }

    cout << "PDA given as input :\n";
    cout << "States = ";
    for (i = 0; i < states.size(); i++) {
        cout << states[i] << " ";
    }
    cout << endl;
    for (i = 0; i < from_state.size(); i++) {
        cout << "δ(" << from_state[i] << "," << input[i] << "," << top_stack[i] << ") = (" << to_state[i] << "," << push_stack[i] << ")\n";
    }
    breakline();
    cout << "Corresponding CFG :\n";
    PDA_to_CFG(from_state, input, top_stack, to_state, push_stack, states);

    return 0;
}

File: pda_to_cfg.js
// Function to generate permutations
function generatePermutations(states, temp, permutes, k) {
  if (k === 0) {
      permutes.push([...temp]);
  } else {
      for (let i = 0; i < states.length; i++) {
          temp[k - 1] = states[i];
          generatePermutations(states, temp, permutes, k - 1);
      }
  }
}

// Function to print permutations
function printPermutations(fromState, input, topStack, toState, pushStack, states) {
  let i, j;
  const temp = new Array(pushStack.length).fill('');
  const permutes = [];

  generatePermutations(states, temp, permutes, pushStack.length);

  for (i = 0; i < permutes.length; i++) {
      let output = `[${fromState}${topStack}${permutes[i][0]}] -> ${input} [${toState}${pushStack[0]}`;
      for (j = 1; j < permutes[i].length; j++) {
          output += `][${permutes[i][j]}${pushStack[j]}`;
      }
      output += `][${permutes[i][0]}]\n`;
      console.log(output);
  }
}

// Function to convert PDA to CFG
function PDA_to_CFG(fromState, input, topStack, toState, pushStack, states) {
  let i, j;
  const n = fromState.length;

  console.log(`Start State = S -> [${states[0]}Z${states[0]}]`);
  for (i = 1; i < states.length; i++) {
      console.log(` | [${states[0]}Z${states[i]}]`);
  }

  for (i = 0; i < n; i++) {
      console.log(`\nδ(${fromState[i]},${input[i]},${topStack[i]}) = (${toState[i]},${pushStack[i]})`);
      if (pushStack[i] === "ε") {
          console.log(`[${fromState[i]}${topStack[i]}${toState[i]}] -> ${input[i]}`);
      } else {
          printPermutations(fromState[i], input[i], topStack[i], toState[i], pushStack[i], states);
      }
  }

  console.log();
  for (i = 0; i < states.length; i++) {
      for (j = 0; j < states.length; j++) {
          console.log(`[${states[i]}Z${states[j]}] -> ε`);
      }
  }
}

// Main function
function main() {
  const fromState = [];
  const toState = [];
  const topStack = [];
  const input = [];
  const pushStack = [];
  const states = [];

  const readline = require('readline').createInterface({
      input: process.stdin,
      output: process.stdout
  });

  readline.question("Enter states of the PDA separated by spaces: ", (line) => {
      states.push(...line.trim().split(' '));

      console.log("Enter PDA transitions in the format δ(q,0,Z) = (q,XZ):");
      console.log("(Press Enter after each transition, type 'done' to finish)");

      const inputTransitions = [];

      function processTransitions() {
          if (inputTransitions.length === 0) {
              console.log("No transitions provided. Exiting.");
              readline.close();
              return;
          }

          inputTransitions.forEach((transition) => {
              const [from, inp, stackTop, toPush] = transition.split(',').map(str => str.trim());
              fromState.push(from);
              input.push(inp);
              topStack.push(stackTop);
              toState.push(toPush.split(',')[0]);
              pushStack.push(toPush.split(',')[1]);
          });

          console.log("PDA given as input:");
          console.log("States =", states.join(' '));
          for (let i = 0; i < fromState.length; i++) {
              console.log(`δ(${fromState[i]}, ${input[i]}, ${topStack[i]}) = (${toState[i]}, ${pushStack[i]})`);
          }

          console.log("Corresponding CFG:");
          PDA_to_CFG(fromState, input, topStack, toState, pushStack, states);
          readline.close();
      }

      (function readTransitions() {
          readline.question("Transition: ", (line) => {
              if (line === "done") {
                  processTransitions();
              } else {
                  inputTransitions.push(line);
                  readTransitions();
              }
          });
      })();
  });
}

// Run the main function
main();
